### DISCUSSION TITLE
Add a decorator syntax for registering viewsets with routers

### DISCUSSION THREAD

**Discussion by daggaz** [encode/django-rest-framework#8917] [Ideas & Suggestions] [answered]
I've always thought it would be good to allow a decorator style syntax for registering routes:

Currently, I always have a list of viewset classes, and at the end of the file a list of registrations for those classes.

Instead, something like this would be nice:
```python
@router.register('some-path')
class MyViewSet(viewsets.ViewSet):
    ...
```

The only change required would be to make the `viewset` parameter to `register()` a keyword argument (default `None`), and if it is not passed, return a decorator instead.

**Comment by daggaz** [comment]
I created a quick PR implementing this:
https://github.com/encode/django-rest-framework/pull/8918

**Comment by auvipy** [comment]
the syntax looks small n good, but we have to reach a consensus from all the old maintainers of @encode/django-rest-framework as we are very careful introducing new API surface to the framework at this stage of projects life cycle.

**Reply by daggaz** [reply]
Yes, the nice thing about it, is that it doesn't alter any existing usage. Previously calling `register()` without a second argument would have resulted in a `TypeError`.

**Comment by baseplate-admin** [comment]
To be honest i dont think adding decorator based routing is a good idea for this project. ( as it will add more churn to an already established library that is used by thousands of projects | Plus not everyone is a fan of this type of routing )

If you want decorator based routing maybe you can look into [`django-ninja`](https://github.com/vitalik/django-ninja). Which implements FastAPI's api

**Reply by carltongibson** [reply, marked as answer]
I'm not a fan. Route resolution becomes dependent on import order.
